BigQuery UDFs (#246) and array literals (#337)

barrywhart opened the pull request:
Addresses #246 and #337

- [x] Add support for complex parameter types, e.g. `ARRAY<STRUCT<product_id INT64, rating FLOAT64>>`
- [x] Add support for `[OPTIONS (library = library_array)]`
- [x] Test this PR against our real-life projects' UDFs
```

barrywhart commented:
@alanmcruickshank: I have a couple of questions/requests for help with this PR. Other than these two areas, the PR is pretty close to complete (probably needs a few more `.sql` examples, not much else).

Thanks for any help you can offer! Once this PR and #320 are in place, I believe SQLFluff will be able to parse very close to 100% of the SQL code in the half dozen or so projects I've tested it on. 🎉 

These are my questions:
1. For non-trivial JavaScript UDFs (more than one line), I need to update the lexer to support triple-quoted strings. I don't know how to do that. Can you help?
2. I've run into a tricky grammar/parsing issue and don't know how to get past it. It involves `STRUCT` types with more than one field. Can you take a look and make any suggestions how to get this working?

Here's more detail about the `STRUCT` issue...

JavaScript UDFs support aggregate types (i.e. `ARRAY` and `STRUCT`), whose syntax is similar to Java generic types or STL containers in C++.

Examples:
```
CREATE TEMP FUNCTION
  moveRedundantToEnd(originalList ARRAY<STRING>, pastActionVals ARRAY<INT64>, pastActionRecs ARRAY<STRING>)
  RETURNS string
  LANGUAGE js AS " <code goes here> "
```

```
CREATE TEMP FUNCTION parseRecArr(arr_str STRING)
RETURNS ARRAY<STRUCT<product_id INT64, rating FLOAT64>>
LANGUAGE js AS " <code goes here > "
```

An `ARRAY` contains unnamed elements of a single type, e.g. `ARRAY<STRING>`. This seems to be parsing okay.

A `STRUCT` contains named fields of potentially disparate types, e.g. `STRUCT<product_id INT64, user_id INT64>`.

I have the basic grammar in place to support both of these. For some reason, I see that a `STRUCT` with a single field parses okay, e.g.:
```
CREATE FUNCTION
qs(
    y STRUCT<product_id INT64>
) LANGUAGE js AS " <code goes here > "
```

but a `STRUCT` with more than one field fails to parse:
```
CREATE FUNCTION
qs(
    y STRUCT<product_id INT64, user_id INT64>
) LANGUAGE js AS " <code goes here > "
```

From tinkering with the grammar, I am _guessing_ that there's an issue with the commas in a multi-field `STRUCT`, perhaps because the comma-separated list of `STRUCT` fields is essentially "nested" inside the comma-separated list of function parameters. I've seen places in the grammar where you use both a `match_grammar` and a `parse_grammar`. Is something like that needed here?

The area of the grammar where this is occurring is named `UDFParameterTypeSegment`. If you have any ideas, that would be 💯.

alanmcruickshank commented:
@barrywhart - I think the bit on triple quoted strings is done. I addressed that in #170. Still worth checking that it works as expected though!

As for the second point on `STRUCT` types, looking at the structure, I *think* it might be more efficient to do it with a modified version of the `Bracketed` grammar, because it deals with the nested nature of those end markers.

I'll see if I can find some time this afternoon to show you what I mean.

barrywhart commented:
@alanmcruickshank: Can you point me to the area of #170 you're referring to? To clarify, when I say "triple quoted strings", I mean something like this example (the part after `LANGUAGE js AS`):

```
CREATE TEMP FUNCTION
  indicesSortDESC(Q ARRAY<float64>,
         cols ARRAY<string>)
  RETURNS string
  LANGUAGE js AS """
    var vals_and_inds = Q.map((value, index) => [value, index]).sort((a, b) => (b[0] - a[0]));
    var inds_only = vals_and_inds.map((value, index)=>(value[1]));
    let recs = []
    for(let i=0;i<inds_only.length;i++)
    {
      recs.push(cols[inds_only[i]]);
    }
    return recs;
    """;
```

In #170, I see support for `LIKE` queries with _single_ quoted strings, e.g.:
```
SELECT *
FROM test
WHERE name NOT LIKE '%y';
```

I tried a quick experiment with using `Bracketed`. Is this what you have in mind?

Changed the `Bracketed` constructor:
```
    def __init__(self, *args, **kwargs):
        self.square = kwargs.pop('square', False)
        self.angle = kwargs.pop('angle', False)
        # Start and end tokens
        # The details on how to match a bracket are stored in the dialect
        if self.square:
            self.start_bracket = Ref('StartSquareBracketSegment')
            self.end_bracket = Ref('EndSquareBracketSegment')
        elif self.angle:
            self.start_bracket = Ref('LessThanSegment')
            self.end_bracket = Ref('GreaterThanSegment')
        else:
            self.start_bracket = Ref('StartBracketSegment')
            self.end_bracket = Ref('EndBracketSegment')
```

Changed the end of the `UDFParameterTypeSegment` grammar to read:
```
        Sequence(
            'STRUCT',
            Bracketed(
                Delimited(  # Comma-separated list of field names/types
                    Sequence(
                        Ref('ParameterNameSegment'),
                        Ref('UDFParameterTypeSegment')
                    ),
                    delimiter=Ref('CommaSegment')
                ),
                angle=True
            ),
        ),
```

With this change, I still see a parse error with a multi-field `STRUCT`:

Input:
```
CREATE FUNCTION
qs(
    y STRUCT<product_id INT64, user_id INT64>
) LANGUAGE js AS " <code goes here > "
```

Output:
```
sqlfluff parse --dialect bigquery test.sql
=== [test.sql] ===
Traceback (most recent call last):
  File "/Users/bhart/.pyenv/versions/sqlfluff-3.6.4/bin/sqlfluff", line 11, in <module>
    load_entry_point('sqlfluff', 'console_scripts', 'sqlfluff')()
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/cli/commands.py", line 338, in parse
    for parsed, violations, time_dict, f_cfg in result:
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 909, in parse_path
    recurse=recurse, config=config),
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 634, in parse_string
    for unparsable in parsed.iter_unparsables():
AttributeError: 'NoneType' object has no attribute 'iter_unparsables'
```

barrywhart commented:
I think I got the modified `Bracketed` working. I also found and updated `BaseGrammar._bracket_sensitive_look_ahead_match()`

```
        start_brackets = [
            parse_context.dialect.ref('StartBracketSegment'),
            parse_context.dialect.ref('StartSquareBracketSegment'),
            parse_context.dialect.ref('LessThanSegment')
        ]
        end_brackets = [
            parse_context.dialect.ref('EndBracketSegment'),
            parse_context.dialect.ref('EndSquareBracketSegment'),
            parse_context.dialect.ref('GreaterThanSegment')
        ]
```

I'd better run all the tests to see if this breaks anything...

barrywhart commented:
@alanmcruickshank: Okay, latest status (sorry for the large number of messages!):
* I ran a test where the UDF used multi-line, triple quoted strings, and it just worked™️. I thought I'd have to reference a different part of the grammar to support both single and triple quotes. Nice!
* I ran all the tests, and my hacked-up `Bracketed` is breaking several other tests. I'm not surprised this is tricky, since "angle brackets" are also the "greater than" and "less than" comparison operators. I appreciate any guidance you can offer. There's no big rush on this; this PR is just the next step in the goal of long-term BigQuery compatibility. Thanks for all your help!

alanmcruickshank commented:
Can you push your changes to `Bracketed` up onto this branch? I'll take a look. I *think* I built the grammar with this kind of case in mind, but I might have not done it very well.

I *think* that you only want to include the angle brackets in `_bracket_sensitive_look_ahead_match` when you're in your angled bracketed segments (i.e. when `Bracketed(angle=True)`). I think that will mean extending `_bracket_sensitive_look_ahead_match` to allow you to pass in arbitrary _extra bracket-like segments_ for that call. If we don't make that configurable then the following code will raise a exception for not having a closing bracket:

```
SELECT 1 < 6 as some_column FROM blah
```

Because the `<` will always be interpreted as potential bracket counting thing.

Additionally I don't think we should use `LessThanSegment` and `GreaterThanSegment` as the elements we match on. I think we should have two *new* segments with similar construction, but with different `name` and `type` parameters so that in the parsed file structure, they're interpreted as brackets and not as operators.

If those rambling make any sense than hopefully they're useful - if not them I'm totally happy to collaborate on this one. :)

barrywhart commented:
Makes sense (I think)! I'll reconstruct those changes and push them up today or tomorrow. I'll see if I can try and apply your other suggestions as well. If I run into trouble, I'm 👍 for collaborating.

barrywhart commented:
@alanmcruickshank: I made the changes you suggested. Sadly, I seem to have introduced _more_ issues now. It's certainly possible I misunderstood your intent or implemented something incorrectly.

Anyway, these are the two most recent commits with the experimental changes to `Bracketed` and `_bracket_sensitive_look_ahead_match`:

```
commit 23a2d571406b4a4b59a7737f923f0ba30848ccef (HEAD -> bhart-issue_246_bigquery_udfs, origin/bhart-issue_246_bigquery_udfs)
Author: Barry Hart <barry.hart@mailchimp.com>
Date:   Wed May 13 18:18:01 2020 -0400

    Add a new parameter to _bracket_sensitive_look_ahead_match: extra_bracket_like_segments

commit 0ebccc182923b5cdcd9735096a0e5584097a594e
Author: Barry Hart <barry.hart@mailchimp.com>
Date:   Wed May 13 17:41:10 2020 -0400

    First non-working attempt at reusing "Bracketed" for angle brackets
```

As I mentioned earlier, the commit prior to these two seemed to work for `ARRAY` and single-field `STRUCT`, but couldn't handle multi-field `STRUCT`s. If you can take a look and help get this back on track, that'd be great. This is a pretty good test SQL:

```
CREATE FUNCTION
qs(
    x ARRAY<INT64>,
    y STRUCT<product_id INT64, user_id INT64>
) LANGUAGE js AS "JavaScript code goes here"
```

alanmcruickshank commented:
Understood. I'm doing a bit a reshuffle now - hold up on any more development on this for a sec, I'll push my changes to this branch. There are a couple of things we can do to make this a bit easier 👍.

barrywhart commented:
Great, thanks!

alanmcruickshank commented:
@barrywhart - I haven't squashed any commits yet, but I'd love your feedback on where I got to. In particular check out the test cases I've added which hopefully cover the use cases you had in mind, specifically the nested datatypes too.

- I've actually done a bit of a deeper refactor of the `Bracketed` grammar to use the new sets functionality to make brackets for dialects a bit more configurable.
- I've also seperated the BigQuery specific things out from the more generic bits and got some level of Postgres specific functions working along with the Bigquery ones.
- There's some more tidying in there and using a few other common elements too to improve modularity.

barrywhart commented:
I'll take a look soon. I glanced over bits of it last night. Thanks so much for all the effort you've put into moving this ahead. It's great for me, and looks like some good architectural progress for future work as well. 👍💯

barrywhart commented:
I did some testing. All the test cases I had provided so far are working.

I tried a more complex test -- a structure containing an array, and this fails to parse. (To be fair, it's entirely possible my original attempt at the grammar did not support this.)

```
CREATE TEMP FUNCTION
CREATE TEMP FUNCTION
qs(
    foo2 INT64
)
RETURNS STRUCT<product_id ARRAY<INT64>>
LANGUAGE js AS """
    return {"product_id": [foo2, foo2 * 2]}; 
"""
```

I haven't had time to look at the new structure of the grammar, so this may very well be something I can fix myself, but I wanted to mention it. I'd say support for this is optional for a first version, anyway -- I don't think any of Mailchimp's code uses this.

I did try and run the above in BigQuery to confirm it is supported there. (I tested it by calling:
```
SELECT qs(1)
```

and it returned:
```
[1, 2]
```

barrywhart commented:
@alanmcruickshank: Ok, I confirm that the core parsing issues with angle brackets and commas have been fixed by your refactoring. 🎉 🎈 

In testing this, I noticed a slight regression, in that  array types (e.g. `ARRAY<INT64>` were no longer working. I narrowed this down to `DatatypeSegment` in `src/sqlfluff/dialects/dialect_bigquery.py`. As an experiment, I replaced its `match_grammar` with an updated version of my earlier `UDFParameterTypeSegment` grammar. With that one change, it seems everything is 💯. In order to convince myself, I verified we can now parse this monstrous-looking function:

```
CREATE TEMP FUNCTION
qs(
    foo1 INT64,
    foo2 ARRAY<STRING>,
    foo3 STRUCT<x INT64>,
    foo4 STRUCT<x INT64, y INT64>,
    foo5 STRUCT<a ARRAY<FLOAT>, b STRUCT<x INT64, y INT64>>
)
RETURNS STRUCT<product_id ARRAY<INT64>>
LANGUAGE js AS """
    CODE GOES HERE
"""
```

I added the above as a new test `.sql` file, 
`test/fixtures/parser/bigquery/create_js_function_complex_types.sql`.

I have a few more things to do, but I think the remaining things will be easy.

I'm really happy with the progress this week. Thanks *SO MUCH* for your help!

barrywhart commented on src/sqlfluff/dialects/dialect_bigquery.py:
```diff
@@ -114,4 +122,193 @@ class IntervalExpressionSegment(BaseSegment):
+                    optional=True
+                )
+            ),
+            'WINDOW',
```
What is this stuff on lines 148-159? I don't recognize it as anything in BigQuery.

alanmcruickshank replied:
They're copies of some of the syntax for postgres. I had just assumed that they were still options in Bigquery - my bad. I'll remove them.

barrywhart commented:
@alanmcruickshank: Ok, this PR is ready for review.

barrywhart commented on src/sqlfluff/dialects/dialect_ansi.py:
```diff
@@ -190,6 +197,19 @@ class IntervalExpressionSegment(BaseSegment):
 
 
+@ansi_dialect.segment()
+class ArrayLiteralSegment(BaseSegment):
```
This addresses a new issue related to UDFs, #337. UDFs can take `ARRAY` parameters, but this is needed in order to allow passing an array literal when calling a function. This syntax is BigQuery specific, but similar to a syntax supported by Postgres. See the linked issue for details.

barrywhart commented on test/fixtures/parser/bigquery/select_where_less_than.sql:
```diff
@@ -0,0 +1,9 @@
+FROM
+  table_a
+WHERE
+  -- Tests that '<' is parsed correctly. (Since some dialects use angle
```
@alanmcruickshank: I found a new parsing error using the BigQuery dialect. This query parses correctly in `master`. I strongly suspect that somehow the new angle bracket handling is causing this. This file fails to parse with the following error message:

```
Couldn't find closing bracket for opening bracket.
```

Any ideas?

alanmcruickshank replied:
Crud - I think I was overoptimistic with something else. Thanks for adding the test case. I've got a good idea about how to resolve it.

dmateusp replied:
I imagine that's resolved now? I just checked out the PR and tried to parse that file and that worked

alanmcruickshank commented:
@barrywhart - I've worked out a workaround to the problem to get it to work. There's a possibility that I've over convoluted things. Would appreciate a bit of a code review on the end result. Especially the changes to `_bracket_sensitive_look_ahead_match`.

barrywhart commented:
I'll take a look!

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -463,24 +463,30 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+        end_brackets = [parse_context.dialect.ref(seg_ref) for seg_ref in end_bracket_refs]
+        start_definite = list(definitely_bracket)
+        end_definite = list(definitely_bracket)
+        # Any any bracket-like things passed as arguments
```
Typo: Should be "Add any"?

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -443,7 +443,7 @@ def _look_ahead_match(cls, segments, matchers, parse_context, code_only=True):
 
     @classmethod
-    def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context, code_only=True):
+    def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context, code_only=True, start_bracket=None, end_bracket=None):
```
I agree this function has become pretty long and complex. If I were to step through it in a debugger, I could probably figure out what it does. Some possible "low-hanging fruit" for simplifying it:
* Extract some helper functions to make the function shorter. 
* There are a lot of local variables in this function.
  * This much state suggests maybe organizing the code into a class, e.g. `BracketMatcher`?  If not that, at minimum I'd add some comments to the most important variables, explaining what they contain and what that's used for. Maybe consider adding type annotations in this area of the code.

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -496,9 +502,13 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+                        if matcher in start_brackets and matcher not in end_brackets:
                             # Same procedure as below in finding brackets.
                             bracket_stack.append(match.matched_segments[0])
+                            definite_stack.append(start_definite[start_brackets.index(matcher)])
```
It looks like `bracket_stack` and `definite_stack` are always updated in unison. It might be easier to understand if you combined them into a single list of dictionaries, objects, whatever.

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -507,33 +517,54 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+                        # bracket. That's only allowed if `not definitely_bracket`.
+
+                        # Any any not definitely brackets
+                        if any(not definite for definite in definite_stack):
```
Could this be replaced with `if not all(definite_stack)`?

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -507,33 +517,54 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+                        # Any any not definitely brackets
+                        if any(not definite for definite in definite_stack):
+                            # If so, remove the last one from the stack and try again.
+                            for idx, elem in enumerate(reversed(definite_stack)):
```
Is this loop safe? It looks like it's removing items from a collection while looping over it. This breaks with most languages and collections I'm familiar with.

This could also be a good candidate for splitting into a separate function.

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -463,24 +463,30 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+        # These are currently strings which need rehydrating
+        start_brackets = [parse_context.dialect.ref(seg_ref) for seg_ref in start_bracket_refs]
+        end_brackets = [parse_context.dialect.ref(seg_ref) for seg_ref in end_bracket_refs]
+        start_definite = list(definitely_bracket)
```
What does `definitely_bracket` contain here? I wasn't sure, so `list(definitely_bracket)` confused me. I had guessed it was either `True` or `False`.

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -463,24 +463,30 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+        # to the list of matchers. We get them from the relevant set on the
+        # dialect. We use zip twice to "unzip" them. We ignore the first
+        # argument because that's just the name.
+        _, start_bracket_refs, end_bracket_refs, definitely_bracket = zip(*parse_context.dialect.sets('bracket_pairs'))
```
This line is pretty tricky to understand. Possible to use `namedtuple`, object, dataclass, or something like that to make the data more "self describing"?

barrywhart commented on src/sqlfluff/parser/grammar.py:
```diff
@@ -507,33 +517,54 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
+                            # If so, remove the last one from the stack and try again.
+                            for idx, elem in enumerate(reversed(definite_stack)):
+                                if not elem:
+                                    del bracket_stack[-idx]
```
This loop and the `del` lines are confusing. Maybe add a comment with an example or brief description of what this code is doing? Could the stack(s) be encapsulated in a class? Is there another way to code these `del` lines, e.g. something like `.pop()`? Could a class like [`deque`](https://docs.python.org/3/library/collections.html#collections.deque) potentially be useful here?

barrywhart commented:
I looked over `_bracket_sensitive_look_ahead_match`. I haven't studied it closely enough to understand what it's doing, so my review is more a set of questions and potential suggestions, some of which may not make sense. Hope this helps!

barrywhart commented:
Darn, I found another query that doesn't parse. It's a very similar issue as the previous one, but in this case, the `WHERE` clause is comparing elements of arrays. I think it's the addition of square brackets causing the issue this time.

See the file `test/fixtures/parser/bigquery/select_where_array_element_less_than.sql`:
```
SELECT
  *
FROM
  `project.dataset.table_1`
WHERE
  effect_size_list[ORDINAL(1)] < effect_size_list[ORDINAL(1+1)]
```

barrywhart commented:
Another one: `test/fixtures/parser/bigquery/select_where_greater_than.sql`:
```
SELECT
  *
FROM
  `project.dataset.actions`
WHERE
  c > 1
```
fails with `Couldn't find closing bracket for opening bracket.`

Just a guesss, but maybe `_bracket_sensitive_look_ahead_match` needs to consider more possibie ways to parse the code? I remember hearing that when C++ added angle brackets for template syntax, it caused parsing challenges, probably similar to this. I wonder if there are any design docs on how they solved it. I'll try and poke around for something...

barrywhart commented:
Some links on how C++ parses angle brackets:
* One response to this [Quora post](https://www.quora.com/How-do-C++-compilers-eliminate-the-need-to-write-white-spaces-between-closing-angle-brackets-when-using-template-classes-since-C++-11-standard) mentions that one C++ parser handles it using a GLR parser, which "is willing to try all potential parses (this makes it much easier to write working grammars)".

alanmcruickshank commented:
Thanks @barrywhart for the feedback, very useful and I think I'll take your advice on splitting out some of that code into helper functions and suchlike. If you haven't already, could you add the failing examples as test cases on this branch?

Also very useful pointers on trying all possible solutions. I had this in mind a while ago, but never got around to implementing it. Sounds like now might be that point!

barrywhart commented:
Another possibility, hackier but perhaps simpler -- is there a way to only allow angle brackets to be considered part of a possible matching pair within the context of a function definition?

I did add a couple of new test SQL files for the above -- I had pasted the SQL in the comments just for ease of reading.

barrywhart commented:
I did some web searches for GLR parsers written in Python. This one looks like it's being actively developed. I suspect you don't plan to make this big a change, but I thought I'd share it. Sometimes I've borrowed ideas and code snippets from packages.

https://pypi.org/project/parglare/

http://www.igordejanovic.net/parglare/stable/

barrywhart commented:
@alanmcruickshank: I think I could contribute at least a few hours a week to this project. Maybe we could chat about how that might work. I would need some support learning the parser, and possibly with occasional tricky issues. Please let me know!

alanmcruickshank commented:
Heya @barrywhart - sorry for the radio silence. I realize this issue floored me a bit and then I lost my flow on the project a little. I'm doing a bit of a reset now and hopefully we'll find a way to get this back on track.

barrywhart commented:
It's totally understandable -- most of us have day jobs and things get crazy sometimes.

If you like, I can take a day or two and deep dive on this myself. With the work you've done already, that could give me some ideas, and I may learn some things that'll let me contribute more in the future. Please let me know how you'd like us to proceed.

alanmcruickshank commented:
Take a look at the vision PR I've just put together. In short I think long term to use something like parglare or another external parsing library makes a lot of sense. It would be a big job to extract, but I think I'd be open to it.

Are there any other suitable candidates for an external parser?

barrywhart commented:
I did a bit more research on GLR parsers. A PyPI search for "GLR" turned up 4 candidates. Below is some info on each project, with the projects sorted from most to least attractive for SQLFluff, based on my initial research. The number after each project is the number of times it was downloaded from PyPI in the past 30 days.

* parglare: 2371. Listed as "beta".  Releases every 2-3 months.
* parsing: 3,554. Listed as "stable", but I can't find the source code repo anywhere. The "Project Home Page" link is dead. 1 release in the past 3 years.
* glrparser: 503. Listed as "alpha". No releases or other activity in the past 2 years.
* jupylr: 133. Project metadata is incomplete. Last release was in 2011.

Do you think it's worthwhile for one or both of us to spend some additional time trying to get this PR to work with the current parser? If you think it _could_ work, I'm inclined to try it. It seems that GLR parsers are less available and less popular, so I'd hate to make a big change only to end up regretting it, possibly for reasons beyond our control (e.g. the parsing framework not being supported anymore).

barrywhart commented:
This PR hit a wall in terms of parser trickiness -- hence the long delay since it was first written. I hope to find time in Q4 to see this through, one way or another. 🤞

dmateusp reviewed (commented):
I was looking at the PR to see if some of the issues it hits can be resolved by `NotExist(...)` that I'm working on in https://github.com/sqlfluff/sqlfluff/pull/640

however, I don't seem to be able to reproduce the issues you talk about in the comments, what is the current status of this PR? Is it blocked by anything?

Also, I just went through a pretty painful rebase on my branch, and I imagine this one will have to go through the same process. The code here seems to be pre-`black` and pre restructuring of the modules.

barrywhart commented:
@dmateusp: I'll take a fresh look -- it's been six months and maybe the issues were fixed by other work. I don't remember the precise issue, but it was clear that the ambiguity was causing one or more of the test queries (and/or other SQLFluff tests) to fail, I think by trying to interpret relational operators as angle brackets. Thanks!

barrywhart commented:
@dmateusp: I ran the tests again in this branch. I'm seeing 4 failures. I'm not sure about the 1st and 4th, but the 2nd and 3rd ones are related to the parsing issues in this PR. Can you please check again and confirm that you're not seeing any test failures?
```
$ py.test -v test/
================================================================================================================== short test summary info ===================================================================================================================
FAILED test/config_test.py::test__config__load_nested - AssertionError: assert {'bar': {'foo...ize': 4}, ...} == {'bar': {'foo...': 'foobar'}}}
FAILED test/dialects_test.py::test__dialect__base_file_parse[bigquery-select_where_array_element_less_than.sql] - sqlfluff.errors.SQLParseError: Couldn't find closing bracket for opening bracket.
FAILED test/dialects_test.py::test__dialect__base_file_parse[bigquery-select_where_greater_than.sql] - sqlfluff.errors.SQLParseError: Found unexpected end bracket!
FAILED test/templaters_test.py::test__templater_full[jinja_f/jinja-True] - assert ('file', (('s... (...))),)),)) == ('file', (('s... (...))),)),))
=============================================================================================================== 4 failed, 418 passed in 10.47s ===============================================================================================================
```

dmateusp commented:
#648 was merged I would assume it would fix some of the issues @barrywhart

barrywhart commented:
Thank you! I'll look at this again today. I believe the PR you mentioned will fix at least one of the 4 failing tests.

barrywhart commented on src/sqlfluff/core/dialects/dialect_bigquery.py:
```diff
@@ -59,11 +60,22 @@
+
+
 # BigQuery allows functions in INTERVAL
+@bigquery_dialect.segment(replace=True)
```
I'm surprised this line wasn't already present. If existing tests weren't breaking, perhaps new tests should be added?

dmateusp replied:
My guess is that there is no test for INTERVAL that uses a function, therefore the ANSI implementation that was being picked up in BigQuery was sufficient

barrywhart commented:
@alanmcruickshank and @dmateusp: Thanks to much help from both of you, I've finally gotten this 7-month old PR (!) synced up with the `master` branch, working, and ready for review. I would much appreciate a review when you have time. (No rush -- after New Year's is fine.)

dmateusp commented on src/sqlfluff/core/dialects/dialect_ansi.py:
```diff
@@ -1766,6 +1786,73 @@ class SetClauseSegment(BaseSegment):
+class CreateFunctionStatementSegment(BaseSegment):
+    """A `CREATE FUNCTION` statement.
+
+    This version in the ANSI dialect should be a "common subset" of the
```
I think we should have it somewhere in the docs, maybe for contributors or inside Architecture, specified that the ANSI dialect is our base dialect and it hosts logic that is common across multiple languages, so it does not strictly represent the ANSI specification

alanmcruickshank replied:
Will make a new issue for this.

dmateusp commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -503,35 +538,52 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
+                            bracket_is_definite = end_definite[
+                                end_brackets.index(matcher)
+                            ]
+                            if bracket_is_definite:
```
I'd like @alanmcruickshank 's stamp of approval here because this bracket logic is new to me, and I added this condition to ignore unexpected end brackets in case the bracket is not definite.

This allows us in BigQuery to reuse `<`, and `>` as brackets without parsing them as different keywords (they're still LowerThan and GreaterThan). And not throw exceptions when they're used as comparison operators.

barrywhart replied:
I agree about getting Alan's review. I think he wrote all the bracket changes for this PR, other than your final changes, Daniel, but at the time, it was more in the spirit of "can we get this working" rather than "is this production quality". I think I mentioned somewhere in this PR, I wonder if the code as written will report bracket-match-related parse errors in a clean and standard way.

dmateusp commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -541,12 +593,18 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
                     )
 
+                    # TODO: Format this better
+                    if definite_bracket_stack:
```
Here it's the same, I'm adding a conditional to relax the closing bracket checks in case the brackets are marked as not definite

dmateusp reviewed (commented):
This is really good 💯 , excited about improved BigQuery support.

I'm going to let @alanmcruickshank give the final approval however, it was my first time looking into the "matching bracket" logic, and I added some comments to the PR to explain the changes I introduced.

alanmcruickshank commented:
Awesome to see this living again - I have to admit, it was this issue that basically burned me out and meant I had to take break from sqlfluff for a few months in the summer. Very glad that it's up and running again - I feel like there might be some closure at hand.

I remember all kinds of Gremlins with this the first time around. I'll make time this week to review it properly. I want to have a good pick through. 🙏

barrywhart commented:
@alanmcruickshank: Aww, I'm sorry. My intuition had told me it might've burned you out, which made me sad because, while it is kind of gnarly technically, this wasn't a blocker or anything for us. I'm very happy you got your energy back. The project seems like it has a great future now, with new contributors and lots of new users, too.

Obviously no rush on reviewing this -- take your time, and if you think it needs improvements before or after merging, I'll do my best to help make that happen. 🙏💯

alanmcruickshank commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -541,12 +593,18 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
-                    raise SQLParseError(
-                        "Couldn't find closing bracket for opening bracket.",
-                        segment=bracket_stack.pop(),
+                    # Check that the unclosed brackets are definite
+                    definite_bracket_stack = list(
+                        filter(lambda b: b[1], zip(bracket_stack, definite_stack))
                     )
 
+                    # TODO: Format this better
+                    if definite_bracket_stack:
```
I think this could be much more concise if we've combined `bracket_stack` and `definite_stack` as per my comment further up. Then this should actually just be:

NB: this is pseudocode, so won't actually work but it gets at the intent and avoids the filter and lambda expressions and makes this more readable.
```python
if any(bracket["definite"] for bracket in bracket_stack):
    raise SQLParseError("...")
```

barrywhart replied:
Done!

alanmcruickshank reviewed (changes requested):
Totally on board with this approach 👍 

This bracket logic is already quite complicated so I've been quite picky in my readability feedback on the code. I think it's important that it's as clean as possible to help more people contribute to it.

Aside from the readability comments, the only one that I think deserves further though is my comment about the heuristic for progressively removing brackets if we misclassify any. I'd love some test case queries which have a good mixture of greater than, less than and angle brackets in all mixed together to stress test this heuristic a little. If it turns out to be good enough to handle all but the most convoluted case then I'm fine as is - but we should still label is as a heuristic in case a future one of us needs to adjust that.

If it's not a sufficient heuristic, then I the overall approach still holds, we just need a more thorough "search" algorithm for working out which of our `<` and `>` characters we mis-classified.

barrywhart commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -475,9 +488,15 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
 
                     if match:
-                        if matcher in start_brackets:
+                        # NB: We can only consider this as a nested bracket if the start
+                        # and end tokens are not the same. If a matcher is both a start and
+                        # end token we cannot deepen the bracket stack.
+                        if matcher in start_brackets and matcher not in end_brackets:
```
@alanmcruickshank: I think you contributed this part of the code. Possibly you were thinking about things like single or double quotes? IIUC, those are handled in a separate part of the parser, and they delineate a single token with "stuff" inside. So... I think I agree with you, but I don't know the parser that well. Do you think this could (should?) be removed? Or I could add a code comment something like this confusing one. 😄

barrywhart replied:
For now, I'm planning to add this comment:
```
                        # NB: We can only consider this as a nested bracket if the start
                        # and end tokens are not the same. If a matcher is both a start and
                        # end token we cannot deepen the bracket stack. In general, quoted
                        # strings are a typical example where the start and end tokens are
                        # the same. Currently, though, quoted strings are handled elsewhere
                        # in the parser, and there are no cases where *this* code has to
                        # handle identical start and end brackets. For now, consider this
                        # a small, speculative investment in a possible future requirement.
```

barrywhart commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -486,14 +505,30 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
+
+                        # Any any not definitely brackets
+                        if any(not definite for definite in definite_stack):
+                            # If so, remove the last one from the stack and try again.
```
I'll add a comment along the above lines, assuming that it's a heuristic for now.

barrywhart commented:
@alanmcruickshank, @dmateusp: I think I've addressed all the feedback suggesting code changes, comments, reorg, etc. 

The remaining item on my list is to add some more thorough SQL tests with relational operators. Do you think it's worth exploring the possibility of dynamically generating these test cases? This _could_ be useful, especially if there's a chance we'll  want to use the same technique in the future. In the past, I've used a property-based testing package called [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) for this. Thoughts?

dmateusp commented:
I had a quick look at Hypothesis and I think it would be great to have an example of it to find edge cases like you say!

barrywhart commented:
@dmateusp: Heh, within 15 minutes of trying Hypothesis, I found a bug, but not in this PR -- I got a `RecursionError` in a query that had 34 `<` operators. I created an issue for it, #669. Hurray for Hypothesis, I guess? I'll keep using. Hypothesis on this PR, but with a lower limit on the query size.

barrywhart commented:
@dmateusp, @alanmcruickshank: How do I add a new dev package to the build? I added a line `hypothesis` to `requirements_dev.txt`, but I don't see it being installed on this line in CircleCI:

```
py37 installed: appdirs==1.4.4,attrs==20.3.0,bench-it==1.0.1,black==20.8b1,cached-property==1.5.2,cdifflib==1.2.5,certifi==2020.12.5,chardet==4.0.0,click==7.1.2,colorama==0.4.4,configparser==5.0.1,coverage==5.3.1,dataclasses==0.6,diff-cover==2.6.1,doc8==0.8.1,docutils==0.16,flake8==3.8.4,flake8-black==0.2.1,flake8-docstrings==1.5.0,idna==2.10,importlib-metadata==3.3.0,inflect==5.0.2,iniconfig==1.1.1,Jinja2==2.11.2,jinja2-pluralize==0.3.0,MarkupSafe==1.1.1,mccabe==0.6.1,mypy==0.790,mypy-extensions==0.4.3,oyaml==1.0,packaging==20.8,pathspec==0.8.1,pbr==5.5.1,pluggy==0.13.1,py==1.10.0,pycodestyle==2.6.0,pydocstyle==5.1.1,pyflakes==2.2.0,Pygments==2.7.3,pyparsing==2.4.7,pytest==6.2.1,pytest-cov==2.10.1,PyYAML==5.3.1,regex==2020.11.13,requests==2.25.1,restructuredtext-lint==1.3.2,six==1.15.0,snowballstemmer==2.0.0,-e git+git@github.com:sqlfluff/sqlfluff.git@b1835231c5d25d6879d7d56b613fd0263979ae41#egg=sqlfluff,stevedore==3.3.0,toml==0.10.2,typed-ast==1.4.1,typing-extensions==3.7.4.3,urllib3==1.26.2,zipp==3.4.0
```

And the test fails because `hypothesis` is not installed.

barrywhart commented:
I figured out the package thing -- `.circleci/config.yml` was missing `requirements_dev.txt` from the cache key. I went ahead and added it -- I hope that was okay.

barrywhart commented:
@alanmcruickshank, @dmateusp: I added a test that autogenerates lots of test scenarios using relational operators. I believe this PR is ready for final review and (hopefully) merge. Thanks!

alanmcruickshank reviewed (approved):
I'm a big fan of hypothesis, I'm really glad we've found a way of getting this into the project.

I think this is ready to rock. Merge at will 👍 .

Thanks so much @barrywhart @dmateusp for bringing this PR back to life.

alanmcruickshank commented:
...assuming you fix the linting errors of course 😄 (but they look easy).

barrywhart commented:
Ok, the linter warnings should be fixed. I'll wait to confirm the build passes, then merge this.

Thanks to both of you, @alanmcruickshank and @dmateusp! Y'all did the tricky bits and I just wrote the grammar and test parts. Onward and upward... 🎉

barrywhart commented:
Merging PR. I see the `dbt017-py38` check failed, but I think that's been seen and is being addressed elsewhere.
